js/web/lib/wasm/jsep/webgpu/ops/attention.ts

```diff
@@ -1004,21 +1004,188 @@ const prepare = (context: ComputeContext, parameters: AttentionParameters) => {
+          }
+        }
+        workgroupBarrier();
+      }
+
+      for (var col = 0; col < ${colsPerThread}; col++) {
+        S_i_j[local_id.y][u32(col * 8) + local_id.x] = acc[col] *  ${type}(uniforms.alpha);
+      }
+
+      workgroupBarrier();
+      let m_i_j_1 = m_i_j;
+      for (var m = 0; m < ${bC}; m++) {
+        m_i_j = max(m_i_j, S_i_j[local_id.y][m]);
+      }
+
+      let exp_j_j_1 = exp(m_i_j_1 - m_i_j);
+      l_i_j *= exp_j_j_1;
+      for (var o = 0; o < ${colsPerThread * numTiles}; o++) {
+        O_i[o] *= exp_j_j_1;
+      }
+
+      for (var tile = 0; tile < ${numTiles}; tile++) {
+        KV_j[local_id.y][local_id.x] = V[kvOffset + local_id.y * uniforms.d + local_id.x + u32(tile * 8)];
+        workgroupBarrier();
+
+        for (var d = 0; d < ${bC}; d++) {
+          let p_i_j = exp(S_i_j[local_id.y][d] - m_i_j);
+          if (tile == 0) {
+            l_i_j += p_i_j;
+          }
+
+          for (var col = 0; col < ${colsPerThread}; col++) {
+            let v_i_j = KV_j[d][(u32(8 * col) + local_id.x) / 4][(u32(8 * col) + local_id.x) % 4];
+            O_i[col * ${numTiles} + tile] += p_i_j * v_i_j;
+          }
+        }
+        workgroupBarrier();
+      }
+    }
+
+    let outputOffset = workgroup_id.z * uniforms.batchOffset + (workgroup_id.x * 32 + local_id.y) * uniforms.headNum
+                       * uniforms.headSize + workgroup_id.y * uniforms.headSize + local_id.x;
+    for (var tile = 0; tile < ${numTiles}; tile++) {
+      for (var col = 0; col < ${colsPerThread}; col++) {
+        let outputIndx = outputOffset + u32(tile * ${bC}) + u32(col * 8);
+        output[outputIndx] = O_i[col * ${numTiles} + tile] / l_i_j;
+      }
+    }
+  }`;
+  };
+  return {
+    name: 'FlashAttentionV2',
+    shaderCache: { hint: `${numTiles}`, inputDependencies },
+    getRunData: () => ({ outputs, dispatchGroup: dispatch, programUniforms }),
+    getShaderSource,
+  };
+};
+
+const applyFlashAttentionV2 = (
+  context: ComputeContext,
+  q: TensorView,
+  k: TensorView,
+  v: TensorView,
+  parameters: AttentionParameters,
+  attributes: AttentionAttrs,
+) => {
+  const inputs = [q, k, v];
+  context.compute(createFlashAttentionV2ProgramInfo(q, k, v, parameters, attributes), { inputs });
+};
+
 export const attention = (context: ComputeContext, attributes: AttentionAttrs): void => {
   const params = validateAttentionInputs(context.inputs, attributes);
 
   const [q, k, v] = prepare(context, params);
 
-  return applyAttention(
-    context,
-    q,
-    k,
-    v,
-    context.inputs[4],
-    undefined,
-    undefined,
-    undefined,
-    context.inputs[5],
-    params,
-  );
+  if (
+    params.sequenceLength >= 1024 &&
+    params.sequenceLength % 32 === 0 &&
+    params.headSize <= 128 &&
+    params.headSize % 32 === 0 &&
+    context.inputs[4] === undefined &&
+    context.inputs[5] === undefined
+  ) {
+    return applyFlashAttentionV2(context, q, k, v, params, attributes);
```

Comment:
Any existing case cover this branch?

Reply:
Currently, I also tested it on sd2.1. From the conditions, we know that the input data size is large, so I am not sure it is reasonable to add an unit test here.